pytest_mh/utils/auditd.py

```diff
@@ -0,0 +1,133 @@
+        :param avc_mode: Action taken when AVC denial is found in audit logs.
+        :type avc_mode: Literal[&quot;fail&quot;, &quot;warn&quot;, &quot;ignore&quot;]
+        :param avc_filter: Regular expression used to filter the AVC denials, defaults to None
+        :type avc_filter: str | None, optional
+        """
+        super().__init__(host)
+
+        self.avc_mode: Literal["fail", "warn", "ignore"] = avc_mode
+        self.avc_filter: str | None = avc_filter
+
+        self.artifacts: set[str] = {"/var/log/audit/audit.log"}
+        self._backup: str | None = None
+
+    def setup(self) -> None:
+        """
+        Create backup of audit logs and clear them for current test run.
+        """
+        super().setup()
+
+        result = self.host.ssh.run(
+            """
+            set -e
+            tmp=`mktemp -d`
+            cp -r --archive /var/log/audit "$tmp"
+            truncate --size 0 /var/log/audit/audit.log*
+            echo $tmp
+            """,
+            log_level=SSHLog.Error,
+        )
+
+        self._backup = result.stdout.strip()
+
+    def teardown(self) -> None:
+        """
+        Restore previous audit logs from backup and remove the backup.
+        """
+        if self._backup is not None:
+            self.host.ssh.run(
+                f"""
+                set -e
+
+                for f in "{self._backup}"/audit/audit.log*; do
+                    name=`basename "$f"`
+                    cat "$f" > "/var/log/audit/$name"
+                done
+
+                rm -fr "{self._backup}"
+                """,
+                log_level=SSHLog.Error,
+            )
+
+        return super().teardown()
+
+    def pytest_report_teststatus(
+        self, report: pytest.CollectReport | pytest.TestReport, config: pytest.Config
+    ) -> tuple[str, str, str | tuple[str, dict[str, bool]]] | None:
+        """
+        Report AVC denial error if found and matches requested filter.
+
+        :param report: Pytest report
+        :type report: pytest.CollectReport | pytest.TestReport
+        :param config: Pytest config
+        :type config: pytest.Config
+        :return: Pytest test status
+        :rtype: tuple[str, str, str | tuple[str, dict[str, bool]]] | None
+        """
+        if report.when != "call":
+            return None
+
+        if self.avc_mode == "ignore":
+            return None
+
+        self.logger.info("Checking for AVC denials")
+
+        result = self.host.ssh.run(
+            "ausearch --input-logs -m AVC,USER_AVC", raise_on_error=False, log_level=SSHLog.Silent
+        )
+        if result.rc:
+            return None
+
+        records = result.stdout
+        if not records:
+            return None
+
+        error_msg = f"AVC denial was detected. Check audit log of {report.nodeid}."
+        if self.avc_filter:
+            match = re.search(self.avc_filter, records)
+            if match is None:
+                return None
+
+            error_msg = (
```

Comment:
It'd be nice to prefix the report with a 'sestatus' and 'getsebool -a | \ on', capturing enabled booleans and policy information.

Reply:
I removed the message since it is not showed anywhere in the latest version.

Anyway, ausearch is able to track any AVC denials, that means also AppArmor. We could add this output elsewhere, maybe preferably to downstream CI as individual step (like gather all package versions, selinux settings, ...)?

Reply:
That's fine, it not going to be looked at often and will probably be more of a nuisance  to see it all the time. ;)